File: scanner.py
from typing import List, Optional

from tokens import SINGLE_TOKENS, KEYWORDS, TokenType, Token


class Scanner:
    def __init__(self, source: str):
        self.source = source
        self.tokens = []
        self.start = 0
        self.current = 0
        self.line = 1
        self.had_error = False

    def is_at_end(self, offset: int = 0) -> bool:
        return (self.current + offset) >= len(self.source)
    
    def advance(self) -> str:
        char = self.source[self.current]
        self.current += 1
        return char
    
    def match_next(self, expected: str) -> bool:
        if self.is_at_end():
            return False
        if self.source[self.current] != expected:
            return False
        self.current += 1
        return True
    
    def peek(self, offset: int = 0) -> str:
        if self.is_at_end(offset):
            return '\0'
        return self.source[self.current + offset]
    
    def scan_string(self):
        while self.peek() != '"' and not self.is_at_end():
            if self.advance() == '\n':
                self.line += 1
        if self.is_at_end():
            print(f'Unterminated string on line {self.line}')
            self.had_error = True
            return
        self.advance()

        value = self.source[self.start+1:self.current-1]
        self.add_token(TokenType.STR, value)

    def scan_number(self):
        while self.peek().isdigit():
            self.advance()
        if self.peek() == '.' and self.peek(1).isdigit():
            self.advance()
            while self.peek().isdigit():
                self.advance()
        value = self.source[self.start:self.current]
        if '.' in value:
            self.add_token(TokenType.FLOAT, value)
        else:
            self.add_token(TokenType.INT, value)

    def scan_identifier(self):
        while self.peek().isalnum() or self.peek() == '_':
            self.advance()
        value = self.source[self.start:self.current]
        ttype = KEYWORDS.get(value)
        if ttype is None:
            ttype = TokenType.IDENT
        self.add_token(ttype)
    
    def add_token(self, ttype: TokenType, literal=None):
        text = self.source[self.start:self.current]
        self.tokens.append(Token(ttype, text, literal, self.line))

    def scan_token(self):
        char = self.advance()

        # match two-character tokens
        if char == '!' and self.match_next('='):
            self.add_token(TokenType.NEQ)
            return
        if char == '>' and self.match_next('='):
            self.add_token(TokenType.GEQ)
            return
        if char == '<' and self.match_next('='):
            self.add_token(TokenType.LEQ)
            return
        if char == ':' and self.match_next('='):
            self.add_token(TokenType.ASSIGN)
            return
        
        # match one-character tokens
        ttype = SINGLE_TOKENS.get(char)
        if ttype is not None:
            self.add_token(ttype)
            return
        
        # handle whitespace
        if char in ' \r\t':
            return
        if char == '\n':
            self.add_token(TokenType.NEWLINE)
            self.line += 1
            return

        # handle literals
        if char == '"':
            self.scan_string()
            return
        if char.isdigit():
            self.scan_number()
            return
        if char.isalpha() or char == '_':
            self.scan_identifier()
            return
        
        print(f'Unexpected character {char} on line {self.line}')
        self.had_error = True

    def scan(self) -> List[Token]:
        while not self.is_at_end():
            self.start = self.current
            self.scan_token()

        self.tokens.append(Token(TokenType.EOF, '', None, self.line))
        return self.tokens

File: src/scanner.rs
use crate::{TokenType, Token};

struct Scanner {
    source: Vec<char>,
    start: usize,
    current: usize,
    line: usize,
}

impl Scanner {
    pub fn new(source: String) -> Self {
        Self {
            source: source.chars().collect(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    fn is_at_end(&self) -> bool {
        self.peek(0) == '\0'
    }

    fn peek(&self, offset: usize) -> char {
        if self.current + offset >= self.source.len() {
            '\0'
        }
        else {
            self.source[self.current + offset]
        }
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.source[self.current - 1]
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.peek(0) != expected {
            return false;
        }
        self.advance();
        true
    }

    fn make_token(&self, ttype: TokenType) -> Token {
        let text: String = self.source[self.start..self.current].iter().collect();
        Token {
            ttype,
            line: self.line,
            text: text,
        }
    }

    fn error_token(&self, message: &'static str) -> Token {
        Token {
            ttype: TokenType::Error,
            line: self.line,
            text: message.to_string(),
        }
    }

    fn handle_comment(&mut self) {
        if !self.match_char('?') {
            return;
        }
        while !self.is_at_end() && self.peek(0) != '\n' {
            self.advance();
        }
    }

    fn handle_comments_and_whitespace(&mut self) {
        while !self.is_at_end() {
            match self.peek(0) {
                '\n' => {
                    self.line += 1;
                },
                '?' => self.handle_comment(),
                x => if !x.is_ascii_whitespace() {
                    break
                },
            };
            if !self.is_at_end() {
                self.advance();
            }
        }
    }

    fn read_string(&mut self) -> Token {
        while !self.is_at_end() && self.peek(0) != '"' {
            if self.peek(0) == '\n' {
                self.line += 1;
            }
            self.advance();
        }
        if self.is_at_end() {
            return self.error_token("Unterminated string");
        }
        self.advance();
        self.make_token(TokenType::Str)
    }

    fn read_number(&mut self) -> Token {
        while self.peek(0).is_ascii_digit() {
            self.advance();
        }
        if self.peek(0) == '.' && self.peek(1).is_ascii_digit() {
            self.advance();
            while self.peek(0).is_ascii_digit() {
                self.advance();
            }
            return self.make_token(TokenType::Float);
        }
        self.make_token(TokenType::Int)
    }

    fn read_ident_or_keyword(&mut self) -> Token {
        while self.peek(0).is_alphanumeric() || self.peek(0) == '_' {
            self.advance();
        }
        let text: String = self.source[self.start..self.current].iter().collect();

        Token {
            ttype: TokenType::keyword_or_ident(text.as_str()),
            line: self.line,
            text,
        }
    }

    fn scan_token(&mut self) -> Token {
        self.handle_comments_and_whitespace();

        self.start = self.current;

        if self.is_at_end() {
            return self.make_token(TokenType::EoF);
        }

        // match single-character tokens
        let c = self.advance();
        if let Some(ttype) = TokenType::single_char_keyword(c) {
            return self.make_token(ttype);
        }
        match c {
            '!' => {
                let is_neq = self.match_char('=');
                return self.make_token(
                    if is_neq {
                        TokenType::NEq
                    }
                    else {
                        TokenType::Bang
                    }
                );
            },
            '>' => {
                let is_geq = self.match_char('=');
                return self.make_token(
                    if is_geq {
                        TokenType::GEq
                    }
                    else {
                        TokenType::GT
                    }
                );
            },
            '<' => {
                let is_leq = self.match_char('=');
                return self.make_token(
                    if is_leq {
                        TokenType::LEq
                    }
                    else {
                        TokenType::LT
                    }
                );
            },
            ':' => {
                let is_assign = self.match_char('=');
                return self.make_token(
                    if is_assign {
                        TokenType::Assign
                    }
                    else {
                        TokenType::Colon
                    }
                );
            },
            '-' => {
                let is_right_arrow = self.match_char('>');
                return self.make_token(
                    if is_right_arrow {
                        TokenType::RightArrow
                    }
                    else {
                        TokenType::Minus
                    }
                )
            },
            _ => (),
        };

        // handle literals
        if c == '"' {
            return self.read_string();
        }
        if c.is_numeric() {
            return self.read_number();
        }
        if c.is_alphabetic() || c == '_' {
            return self.read_ident_or_keyword();
        }

        self.error_token("Unexpected character")
    }

    pub fn scan(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();
        #[cfg(feature = "debug")]
        let mut line = 0;
        loop {
            let token = self.scan_token();
            #[cfg(feature = "debug")]
            {
                if token.line != line {
                    line = token.line;
                    print!("{:04}: ", line);
                }
                else {
                    print!("    | ");
                }
                println!("{:?}", token);
            }
            if token.ttype == TokenType::EoF {
                tokens.push(token);
                return tokens;
            }
            tokens.push(token);
        }
    }
}

pub fn scan(source: String) -> Vec<Token> {
    let mut scanner = Scanner::new(source);
    scanner.scan()
}
